### PR TITLE
Implement Thrift connection pooling

### PR DESCRIPTION
**Pull request by kokosing**
Implement Thrift connection pooling

### CODE REVIEW

**Comment by kokosing** [comment]
Locally I got:
```
20/03/03 17:21:50 [pool-5-thread-7]: DEBUG metastore.RetryingHMSHandler: PERFLOG 7: function=get_table, elapsedTime=183ms, retryCount=0, error=false
20/03/03 17:21:50 [pool-5-thread-7]: DEBUG metastore.RetryingHMSHandler: PERFLOG 7: function=get_all_databases, elapsedTime=16ms, retryCount=0, error=false
20/03/03 17:21:56 [pool-5-thread-7]: DEBUG metastore.RetryingHMSHandler: PERFLOG 7: function=get_table, elapsedTime=58ms, retryCount=0, error=false
20/03/03 17:21:56 [pool-5-thread-7]: DEBUG metastore.RetryingHMSHandler: PERFLOG 7: function=get_all_databases, elapsedTime=10ms, retryCount=0, error=false
20/03/03 17:22:23 [pool-5-thread-7]: DEBUG metastore.RetryingHMSHandler: PERFLOG 7: function=get_table, elapsedTime=55ms, retryCount=0, error=false
```
Where first call for `get_table` was creating the thrift connection session. So it dropped from 158ms -> 58ms in subsequent call.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/ObjectPool.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        }
+    }
+
+    private PoolQueue<T> getQueue(K key)
+    {
+        try {
+            return poolCache.get(key);
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e);
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/ObjectPool.java:80]: Rethrow the cause as `UncheckedExecutionException`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/ObjectPool.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    @Override
+    public void close()
+    {
+        poolCache.invalidateAll();
+    }
+
+    private static class PoolQueue<T extends Closeable>
+            implements Closeable
+    {
+        private final Queue<T> queue = new ArrayDeque<>();
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/ObjectPool.java:101]: That correct but as long spike dies and after `.expireAfterAccess(ttl.toMillis(), TimeUnit.MILLISECONDS)` all of these will be closed.

**Comment by kokosing** [comment]
Tested once again with standalone product test environment simple query: `select * from region r, nation n where r.regionkey = n.regionkey;`

- pooling enabled:

Query runtime is from about

first time
```
presto-master       | 2020-07-21T17:54:03.056+0545      DEBUG   Query-20200721_120904_00021_j3f4u-353   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table(dbname='default', tbl_name='region') succeeded in 134.97ms
presto-master       | 2020-07-21T17:54:03.095+0545      DEBUG   Query-20200721_120904_00021_j3f4u-353   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table(dbname='default', tbl_name='nation') succeeded in 36.78ms
presto-master       | 2020-07-21T17:54:03.167+0545      DEBUG   Query-20200721_120904_00021_j3f4u-355   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table_statistics_req(request=TableStatsRequest(dbName:default, tblName:region, colNames:[regionkey, name, comment])) succeeded in 45.66ms
presto-master       | 2020-07-21T17:54:03.207+0545      DEBUG   Query-20200721_120904_00021_j3f4u-355   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table_statistics_req(request=TableStatsRequest(dbName:default, tblName:nation, colNames:[nationkey, name, regionkey, comment])) succeeded in 38.29ms
presto-master       | 2020-07-21T17:54:03.359+0545      INFO    dispatcher-query-12     io.prestosql.event.QueryMonitor TIMELINE: Query 20200721_120904_00021_j3f4u :: Transaction:[23058e44-8379-4a31-8c9b-a014bc57216d] :: elapsed 434ms :: planning 139ms :: waiting 181ms :: scheduling 205ms :: running 19ms :: finishing 71ms :: begin 2020-07-21T17:54:02.918+05:45 :: end 2020-07-21T17:54:03.352+05:45
```

second time (connection in the pool) (about 100ms faster)
```
presto-master       | 2020-07-21T17:54:09.059+0545      DEBUG   Query-20200721_120910_00022_j3f4u-355   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table(dbname='default', tbl_name='region') succeeded in 34.83ms
presto-master       | 2020-07-21T17:54:09.097+0545      DEBUG   Query-20200721_120910_00022_j3f4u-355   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table(dbname='default', tbl_name='nation') succeeded in 36.23ms
presto-master       | 2020-07-21T17:54:09.161+0545      DEBUG   Query-20200721_120910_00022_j3f4u-353   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table_statistics_req(request=TableStatsRequest(dbName:default, tblName:region, colNames:[regionkey, name, comment])) succeeded in 40.64ms
presto-master       | 2020-07-21T17:54:09.202+0545      DEBUG   Query-20200721_120910_00022_j3f4u-353   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table_statistics_req(request=TableStatsRequest(dbName:default, tblName:nation, colNames:[nationkey, name, regionkey, comment])) succeeded in 39.35ms
presto-master       | 2020-07-21T17:54:09.413+0545      INFO    dispatcher-query-14     io.prestosql.event.QueryMonitor TIMELINE: Query 20200721_120910_00022_j3f4u :: Transaction:[abb339a3-e126-4de2-955d-64f195afa536] :: elapsed 384ms :: planning 145ms :: waiting 79ms :: scheduling 93ms :: running 103ms :: finishing 43ms :: begin 2020-07-21T17:54:09.023+05:45 :: end 2020-07-21T17:54:09.407+05:45
```

* without pooling
```
presto-master       | 2020-07-21T18:01:36.744+0545      DEBUG   Query-20200721_121636_00004_8hvr4-164   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table(dbname='default', tbl_name='region') succeeded in 132.22ms
presto-master       | 2020-07-21T18:01:36.865+0545      DEBUG   Query-20200721_121636_00004_8hvr4-164   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table(dbname='default', tbl_name='nation') succeeded in 118.67ms
presto-master       | 2020-07-21T18:01:37.031+0545      DEBUG   Query-20200721_121636_00004_8hvr4-165   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table_statistics_req(request=TableStatsRequest(dbName:default, tblName:region, colNames:[regionkey, name, comment])) succeeded in 128.29ms
presto-master       | 2020-07-21T18:01:37.175+0545      DEBUG   Query-20200721_121636_00004_8hvr4-165   io.prestosql.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient     Invocation of get_table_statistics_req(request=TableStatsRequest(dbName:default, tblName:nation, colNames:[nationkey, name, regionkey, comment])) succeeded in 140.08ms
presto-master       | 2020-07-21T18:01:37.369+0545      INFO    dispatcher-query-3      io.prestosql.event.QueryMonitor TIMELINE: Query 20200721_121636_00004_8hvr4 :: Transaction:[7575376b-d287-4011-baed-05c106656365] :: elapsed 745ms :: planning 343ms :: waiting 259ms :: scheduling 281ms :: running 37ms :: finishing 84ms :: begin 2020-07-21T18:01:36.609+05:45 :: end 2020-07-21T18:01:37.354+05:45
```

Pooling made this query about two times faster. From about 700ms to 300ms.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/ObjectPool.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/ObjectPool.java:102]: Also TTL is here hardcoded to 30s, I don't want to make it configurable.

File: presto-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/presto/etc/log.properties [hunk 3]
```diff
@@ -1 +1,2 @@
 com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory=WARN
+io.prestosql.plugin.hive=DEBUG
```
Comment (kokosing) [inline presto-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/presto/etc/log.properties:2]: undo

**Comment by kokosing** [comment]
@electrum @dain please take a look at your first convenience

**Comment by losipiuk** [comment]
> Pooling made this query about two times faster. From about 700ms to 300ms.

Just to confirm. You run query multiple times, and variance is small, right?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreModule.java [hunk 4]
```diff
@@ -72,5 +73,20 @@ protected void setup(Binder binder)
...
+                this::usePooledHiveMetastoreClientFactory));
+    }
+
+    private void usePooledHiveMetastoreClientFactory(Binder binder)
+    {
+        newOptionalBinder(binder, ThriftMetastoreClientFactory.class)
+                .setBinding().to(PooledHiveMetastoreClientFactory.class).in(Scopes.SINGLETON);
+        binder.bind(ThriftMetastoreClientFactory.class)
+                .annotatedWith(PooledHiveMetastoreClientFactory.ForPooledMetastore.class)
+                .to(DefaultThriftMetastoreClientFactory.class)
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreModule.java:89]: Using `DefaultThriftClientFactory` here makes it not possible to easily replace implementation via `OptionalBinder` defined above.

Maybe instead do sth like this:

```java
    @Retention(RUNTIME)
    @Target({FIELD, PARAMETER, METHOD})
    @Qualifier
    public @interface Delegate {}

    @Override
    protected void setup(Binder binder)
    {
        newOptionalBinder(binder, Key.get(ThriftMetastoreClientFactory.class, Delegate.class))
                .setDefault().to(DefaultThriftMetastoreClientFactory.class).in(Scopes.SINGLETON);

        install(installModuleIf(
                ThriftMetastoreConfig.class,
                thriftMetastoreConfig -> !thriftMetastoreConfig.isConnectionPoolEnabled(),
                this::useNonPooledHiveMetastoreClientFactory));

        install(installModuleIf(
                ThriftMetastoreConfig.class,
                thriftMetastoreConfig -> thriftMetastoreConfig.isConnectionPoolEnabled(),
                this::usePooledHiveMetastoreClientFactory));
   ....
   }


    private void useNonPooledHiveMetastoreClientFactory(Binder binder)
    {
        binder.bind(ThriftMetastoreClientFactory.class)
                .to(Key.get(ThriftMetastoreClientFactory.class, Delegate.class))
                .in(Scopes.SINGLETON);
    }

    private void usePooledHiveMetastoreClientFactory(Binder binder)
    {
        newOptionalBinder(binder, ThriftMetastoreClientFactory.class)
                .setBinding().to(PooledHiveMetastoreClientFactory.class).in(Scopes.SINGLETON);
        binder.bind(ThriftMetastoreClientFactory.class)
                .annotatedWith(PooledHiveMetastoreClientFactory.ForPooledMetastore.class)
                .to(Key.get(ThriftMetastoreClientFactory.class, Delegate.class))
                .in(Scopes.SINGLETON);
    }
```

Now we should be able to have pooling working with overriden base thrift connection factory by overriding binding on optional binder for `Key.get(ThriftMetastoreClientFactory.class, Delegate.class)`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/PooledHiveMetastoreClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                    }
+                },
+                1000,
+                Duration.valueOf("30s"));
+    }
+
+    @Override
+    public ThriftMetastoreClient create(HostAndPort address, Optional<String> delegationToken)
+    {
+        ObjectPool.Lease<ThriftMetastoreClient> lease = pool.get(new Key(delegationToken, address));
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/PooledHiveMetastoreClientFactory.java:64]: with this implementation we can end up returning a series of broken connections to user which will make retry mechanism in the layer above non-effective.
Example case when we will return a series of broken connections is when we end up with having many connections in the queue in pool after a spike. With limited traffic after the spike some of the connections will time out due to inactivity (probably many at more-or same time).

I think when we got a failure on pooled connection we should return a brand new connection to same caller on next `create` request. I have no concrete proposal how to implement that nicely though.

Alternatively we can add some connection testing mechanism, so connection is verified before being returned to caller. Less bullet proof - but maybe good enough.

**Comment by tooptoop4** [comment]
will this be merged soon?

**Comment by jerryleooo** [comment]
ping

**Review by tooptoop4 — APPROVED** [review]

**Comment by colebow** [comment]
:wave: @kokosing - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**